### ISSUE TITLE
add `from __future__ import annotations` if adding forward-reference annotations

### ISSUE THREAD

**Issue by Jongy** [Instagram/MonkeyType#203]
```
$ cat a.py
class A:
    def f(self):
        return self

$ cat b.py
import a

a.A().f()

$ monkeytype run b.py
$ monkeytype apply a
class A:
    def f(self) -> A:
        return self

$ python b.py
Traceback (most recent call last):
  File "b.py", line 1, in <module>
    import a
  File "a.py", line 1, in <module>
    class A:
  File "a.py", line 2, in A
    def f(self) -> A:
NameError: name 'A' is not defined
```

As described [here](https://www.python.org/dev/peps/pep-0484/#forward-references), I guess forward declarations should have been used here. If I manually use `-> "A"`, then Mypy identifies it correctly.

I see references to `typing.ForwardRef` in MonkeyType's code, so either it was missed for some reason, or am I using it wrong?

MonkeyType version: `20.5.0`.

**Comment by carljm** [comment]
Thanks for the report! The `ForwardRef` references in MonkeyType's code are for cases where we get the existing annotation from the code and need to reproduce it in our stub. MonkeyType doesn't currently attempt to detect forward references in generated stubs and output stringified references. And I don't think it makes sense to add that feature at this point, because it's unnecessary with `from __future__ import annotations`, which in future will be the default behavior of annotations.

So if you aren't using `from __future__ import annotations`, this is just something you need to fix up as needed in the applied annotations. In general MonkeyType doesn't aim to provide ready-for-production code; it aims to provide an informative first draft of type information to be edited as needed by the developer. This also applies to things like generating concrete types when perhaps an abstract type would be more suitable, or unions where a common parent type would be better.

**Comment by Jongy** [comment]
Thanks for the quick reply! I wasn't aware of `from __future__ import annotations`. Like you suggested, I will use it. 

If MonkeyType doesn't attempt to detect forward references at all, perhaps `monkeytype apply` should also insert `from __future__ import annotations`? (just like it inserts any missing import from `typing`)

**Comment by carljm** [comment]
Yeah, that's a reasonable thought. I'm a little hesitant to have MonkeyType do that by default, though, since `from __future__ import annotations` can be a breaking change (if your code is inspecting the annotations at runtime).

**Comment by Jongy** [comment]
Hmm. Then maybe MonkeyType can identify it is about to add a forward reference (if it's easy given the current code) and only if so, inject `from __future__ import annotations`. If done only when required, it is much less of a breaking change.

**Comment by carljm** [comment]
I don't think it's _easy_, but it should be doable. I still have some mixed feelings about MonkeyType doing that, but I think I would accept a PR for it. I'll reopen this issue for that.

**Comment by Jongy** [comment]
> I don't think it's easy, but it should be doable. I still have some mixed feelings about MonkeyType doing that, but I think I would accept a PR for it

Another option is to add a command-line switch to enable this behavior (disregarding "detection of forward references" which indeed may be harder).  Personally, I'm trying to use MonkeyType to add annotations to a medium-sized project, and I'm finding myself manually adding `from __future__ import annotations` quite a few times. I'd happily just pass `--pep-563`.

I can post a PR adding that ^^.

By the way - another issue I found with MonkeyType automatically adding imports, is that they often create circular imports. The common solution I saw online is to use `from typing import TYPE_CHECKING`, and wrap those only-for-annotations `import`s with it. Maybe `import`s we add should be wrapped with it? After all, if they never existed before, they are not needed, and only Mypy cares about them.

I understand your position about MonkeyType not aiming to produce production-ready code, but these 2 issues are just those I noticed to consume most of my time while applying MonkeyType suggestions, so I figured we should improve here :)

**Comment by carljm** [comment]
If you want to use `from __future__ import annotations` in your project, I would suggest just adding it to all files in your project once (this is pretty easy to automate even with just `sed`). Having it in some files and not others makes things confusing for later development that touches annotations.

Regarding automatically putting added imports under `if TYPE_CHECKING:`, that's a great idea, but probably only if we do add `from __future__ import annotations` (since it pretty much means all annotations would need to be stringified otherwise).

I think it would be reasonable to group both of these features under a `--pep-563` flag.

**Comment by Jongy** [comment]
> If you want to use from __future__ import annotations in your project, I would suggest just adding it to all files in your project once (this is pretty easy to automate even with just sed).

Yes, I was indeed planning on doing that. But it's one of those things you tend to forget when adding new files, and it's nicer when the automatic tools do it for your wherever needed :P

I'll take a look on adding both, combined, via `--pep-563`.

**Comment by carljm** [comment]
Closed by #282
